catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,186 @@
+                Integer.MAX_VALUE,
+                indexOffset,
+                "Query request start index must be greater than zero");
+
+        queriedResultsBuffer = new LinkedBlockingDeque<>();
+    }
+
+    // TODO: 6/27/17 Move this method over to QueryResultIterable inner class
+    /**
+     * If there are pages beyond the current page, returns true.
+     *
+     * @return True if there are results beyond the current page
+     */
+    /**
+     * @return true if there are more results for the client
+     */
+    public boolean hasNext() {
+
+        if (notEmpty()) {
+             return true;
+         }
+
+         if (isExhausted) {
+             return false;
+         }
+
+         //Buffer is empty, but query might not be exhausted.
+         fetchNext();
+         return notEmpty();
+    }
+
+    protected boolean notEmpty() {
+        return !queriedResultsBuffer.isEmpty();
+    }
+
+    /**
+     * Catalog Framework queried for results until current page is full of results
+     * or endIndex is reached.
+     *
+     * @return Collection of results
+     * @throws NoSuchElementException if the iteration has no more elements
+     */
+    public List<Result> next() throws NoSuchElementException {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        while (!isExhausted && getBufferSize() < pageSize) {
+            fetchNext();
+        }
+
+        List<Result> queriedResultsBufferCopy = new ArrayList<>(pageSize);
+        queriedResultsBuffer.drainTo(queriedResultsBufferCopy, pageSize);
+        return queriedResultsBufferCopy;
+    }
+
+    /**
+     * Get the next batch of results.
+     * Assume the current index is pointed to the first result
+     * that has not been retrieved.
+     * Update the current index after the results are fetched.
+     * Add the fetched results to the queriedResultsBuffer.
+     * If no results were returned, assume there are no more results possible.
+     */
+    private void fetchNext() {
+
+        queryCopy.setStartIndex(indexOffset);
+        List<Result> resultList = queryCatalogFramework();
+
+        if (resultList.isEmpty()) {
+            isExhausted = true;
+        } else {
+            indexOffset += resultList.size();
+            queriedResultsBuffer.addAll(resultList);
+        }
+    }
+
+    /**
+     * Performs queryCopy of the Catalog Framework with the QueryRequest.
+     *
+     * @return List of results gleaned from CatalogFramework request
+     */
+    private List<Result> queryCatalogFramework() {
+        try {
+            return catalogFramework.query(new QueryRequestImpl(queryCopy))
+                    .getResults();
+        } catch (UnsupportedQueryException | SourceUnavailableException | FederationException e) {
+            throw new CatalogRuntimeException(e);
+        }
+    }
+
+    int getBufferSize() {
+        return queriedResultsBuffer.size();
+    }
+
+    QueryImpl copy(QueryImpl query) {
```

Comment:
❗️ Please reorder methods such that public ones are first, then protected, then package-private, then private.